torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -494,6 +496,256 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+// Create a TensorView before the original tensor. A common use case is to read
+// original tensor into shared memory or local registers.
+// Analogous to TVM Cache_Write
+TensorView* TensorView::cache_before() {
+  FusionGuard fg(this->fusion());
+
+  Expr* origin_expr = this->fusion()->origin(this);
+  TORCH_CHECK(
+      origin_expr != nullptr,
+      "Error adding cache_before ",
+      this,
+      " its origin is a nullptr and we restrict using cache_before on an input.");
+
+  // Create Producer Domain
+  auto dom = domain()->noReductions();
+  std::vector<IterDomain*> out_domain(dom.size(), nullptr);
+  for (size_t i = 0; i < dom.size(); ++i) {
+    if (dom[i]->isBroadcast()) {
+      out_domain[i] = new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true);
+    } else {
+      out_domain[i] = new IterDomain(dom[i]->start(), dom[i]->extent());
+    }
+  }
+
+  // This domain will be the consumer, so create the producer
+  TensorView* producer =
+      new TensorView(new TensorDomain(out_domain), this->getDataType().value());
+
+  // Set domain of consumer
+  TensorView* consumer = this;
+
+  // Setup dependency chain, inserting producer - Cache_Read - before this op.
+  // Before: Prev Op -> This Op
+  // After: Prev Op -> New Cache_Read Op -> This Op
+
+  // Expr* producer_origin =
+  if (origin_expr != nullptr)
+    this->createExprConsumer(origin_expr, producer);
+
+  // Expr* producer_uses =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // ComputeAt - Before
+  if(this->hasComputeAt())
+    producer->computeAt(consumer, getThisComputeAtAxis());
+
+  return producer;
+}
+
+// Create a TensorView after the original tensor. A common use case is to write
+// results into shared memory or local registers before moving to global memory.
+// Analogous to TVM Cache_Read
+TensorView* TensorView::cache_after() {
+  FusionGuard fg(this->fusion());
+
+  // Get all the uses for this Tensorview
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
+  TORCH_CHECK(
+      uses_expr.size() > 0,
```

Comment:
Using cache_after on a tensor shouldn't necessarily be bad if it's on a tensor view without a use as long as it's not an output. I think for now it's restrictive enough to make sure it's not an output.

I'm wondering if there may be uses to do a cache_after and then make the result an output.

Similar for cache_before/inputs.